usvm-ts/src/main/kotlin/org/usvm/TSExprTransformer.kt

```diff
@@ -0,0 +1,149 @@
+package org.usvm
+
+import io.ksmt.sort.KBoolSort
+import io.ksmt.sort.KFp64Sort
+import io.ksmt.utils.cast
+import org.jacodb.ets.base.EtsBooleanType
+import org.jacodb.ets.base.EtsNumberType
+import org.jacodb.ets.base.EtsType
+
+typealias CoerceAction = (UExpr<out USort>, UExpr<out USort>) -> UExpr<out USort>?
+
+class TSExprTransformer(
+    private val baseExpr: UExpr<out USort>,
+    private val scope: TSStepScope,
+) {
+
+    private val exprCache: MutableMap<USort, UExpr<out USort>?> = mutableMapOf(baseExpr.sort to baseExpr)
+    private val ctx = baseExpr.tctx
+
+    init {
+        if (baseExpr.sort == ctx.addressSort) {
+            TSTypeSystem.primitiveTypes.forEach { transform(ctx.typeToSort(it), modifyConstraints = false) }
+        }
+    }
+
+    fun transform(sort: USort, modifyConstraints: Boolean = true): UExpr<out USort>? = with(ctx) {
+        val (result, type) = when (sort) {
+            fp64Sort -> asFp64() to EtsNumberType
+            boolSort -> asBool() to EtsBooleanType
+            addressSort -> asRef() to null
+            else -> error("")
+        }
+
+        if (modifyConstraints && type != null) suggestType(type)
+
+        return result
+    }
+
+    fun intersectWithTypeCoercion(
+        other: TSExprTransformer,
+        action: CoerceAction
+    ): UExpr<out USort> {
+        intersect(other)
+
+        val rawExprs = exprCache.keys.mapNotNull { sort ->
+            val lhv = transform(sort)
+            val rhv = other.transform(sort)
+            if (lhv != null && rhv != null) {
+                action(lhv, rhv)
+            } else null
+        }
+
+        val innerCoercionExprs = this.generateAdditionalExprs(rawExprs) + other.generateAdditionalExprs(rawExprs)
+
+        val exprs = rawExprs + innerCoercionExprs
+
+        return if (exprs.size > 1) {
+            assert(exprs.all { it.sort == ctx.boolSort })
+            UJoinedBoolExpr(ctx, exprs.cast())
+        } else exprs.single()
+    }
+
+    private fun intersect(other: TSExprTransformer) {
+        exprCache.keys.forEach { sort ->
+            other.transform(sort)
+        }
+        other.exprCache.keys.forEach { sort ->
+            transform(sort)
+        }
+    }
+
+    private val addedExprCache: MutableSet<UExpr<out USort>> = mutableSetOf()
+
+    /**
+     * Generates and caches additional constraints for coercion expression list.
+     *
+     * For now used to save link between fp and bool sorts of [baseExpr].
+     *
+     * @return List of additional [UExpr].
+     */
+    private fun generateAdditionalExprs(rawExprs: List<UExpr<out USort>>): List<UExpr<out USort>> = with(ctx) {
+        if (!rawExprs.all { it.sort == boolSort }) return emptyList()
+        val newExpr = when (baseExpr.sort) {
+            addressSort -> addedExprCache.putOrNull(mkEq(fpToBoolSort(asFp64()), asBool()))
```

Comment:
This is unclear, why we add info about bv bool transformation if we have an address

Reply:
Explanation added

Reply:
Why don't we need to add a connection between baseSort expr and them? If it already exists for baseSort <-> fp and bool, it exists for bool and fp by transitivity as well